[draft] Variational Inference

jsalvatier opened the pull request:
Added a draft of what we talked about here: https://github.com/pymc-devs/pymc3/issues/708

Probably needs a bunch of work. The step size on Adagrad doesn't actually go down, which has to be wrong, but as far as I can tell the same is true for the STAN code and the description in the [preprint](http://www.proditus.com/papers/KRGB_preprint.pdf). Also, I'm not averaging together the gradient draws.

jsalvatier committed:
cleanup things

jsalvatier committed:
variational inference draft

jsalvatier committed:
use it on an actual problem

jsalvatier commented:
Let me know if you have thoughts on what the interface should look like.

twiecki commented on pymc3/variational/ADVI.py:
```diff
@@ -0,0 +1,103 @@
+
+def inner_gradients(logp, n, uw):
+    l = uw.size/2
+    u = uw[:l]
```
Are you joining mu and the diagonal into a single vector? Why?

twiecki replied:
Ah, that's probably to give adagrad a single set of params to optimize.

jsalvatier replied:
Yes, I could just do it as two pairs, and in this case it might be easier. I was hoping there would be a nice collection of stochastic gradient optimizers. But there doesn't seem to be so maybe I'll switch.

twiecki replied:
Well it did look like there were other optimizers in lasagne that could be interesting. Not sure how much the authors explored this.

jsalvatier replied:
Oh, I just meant that scipy.optimize for example doesn't have stochastic optimizers and Lasagne didn't have the right one.

jsalvatier committed:
add the init file

akucukelbir commented:
hey guys, just a quick note: dustin tran caught a bug in our Stan code that calculates the adaptive stepsize. it seems like you guys are working off the paper, which should be correct. but just in case you're also working off the Stan code, i recommend looking at the latest commit on feature/bbvb.

jsalvatier commented:
Thanks for the heads up Alp!

On Mon, Jun 22, 2015 at 9:05 AM Alp Kucukelbir notifications@github.com
wrote:

> hey guys, just a quick note: dustin tran caught a bug in our Stan code
> that calculates the adaptive stepsize. it seems like you guys are working
> off the paper, which should be correct. but just in case you're also
> working off the Stan code, i recommend looking at the latest commit on
> feature/bbvb.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/775#issuecomment-114165394.

twiecki commented:
@jsalvatier We probably need to rebase this branch off of master to get the recent speedup changes in. Want me to go ahead and do that?

jsalvatier commented:
Sure!

On Thu, Jun 25, 2015 at 4:51 AM Thomas Wiecki notifications@github.com
wrote:

> @jsalvatier https://github.com/jsalvatier We probably need to rebase
> this branch off of master to get the recent speedup changes in. Want me to
> go ahead and do that?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/775#issuecomment-115220666.

twiecki commented:
OK, should be all cleaned up.

fonnesbeck commented:
The example in the advi notebook fails with a TypeError for me:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-566d165afaa1> in <module>()
----> 1 pymc3.variational.advi( model=model, n=100000)

/Users/fonnescj/GitHub/pymc3/pymc3/variational/ADVI.py in advi(vars, start, model, n)
     27     vars = inputvars(vars)
     28 
---> 29     grad, inarray, shared = variational_gradient_estimate(vars, model)
     30 
     31     for var, share in shared.items():

/Users/fonnescj/GitHub/pymc3/pymc3/variational/ADVI.py in variational_gradient_estimate(vars, model)
     70     l = uw.size/2
     71     n = r.normal(size=(1,))
---> 72     gradient_estimate = inner_gradients(logp, n, uw)
     73 
     74     return gradient_estimate, inarray, shared

/Users/fonnescj/GitHub/pymc3/pymc3/variational/ADVI.py in inner_gradients(logp, n, uw)
     76 def inner_gradients(logp, n, uw):
     77     l = uw.size/2
---> 78     u = uw[:l]
     79     w = uw[l:]
     80 

/usr/local/lib/python3.4/site-packages/theano/tensor/var.py in __getitem__(self, args)
    370             try:
    371                 if arg is not numpy.newaxis:
--> 372                     theano.tensor.subtensor.Subtensor.convert(arg)
    373             except theano.tensor.subtensor.AdvancedIndexingError:
    374                 if advanced:

/usr/local/lib/python3.4/site-packages/theano/tensor/subtensor.py in convert(entry, slice_ok)
    367                 # as slices containing maxsize are not generated by
    368                 # __getslice__ anymore.
--> 369                 slice_b = Subtensor.convert(b, False)
    370             else:
    371                 slice_b = None

/usr/local/lib/python3.4/site-packages/theano/tensor/subtensor.py in convert(entry, slice_ok)
    338                 and (entry.type in invalid_scal_types
    339                      or entry.type in invalid_tensor_types)):
--> 340             raise TypeError("Expected an integer")
    341 
    342         if isinstance(entry, gof.Variable) and entry.type in scal_types:

TypeError: Expected an integer
```

jsalvatier commented:
Ah looks like a problem with `uw.size/2` being a float should probably be `(uw.size/2).astype('int64')` or something

twiecki committed:
Extend advi example.

twiecki commented:
I updated the code a bit as well as the advi example: https://github.com/pymc-devs/pymc3/blob/advi/pymc3/examples/advi.ipynb

One thing I noticed is that the standard deviation returned by `advi` is maybe `log` transformed? CC @jsalvatier

jsalvatier commented:
Yeah, we should exp it before we return it

twiecki commented:
So it's not a bug?

jsalvatier commented:
Not a bug in the main algorithm I think, just a bug in the return value.

twiecki committed:
Exp std estimate from advi.

twiecki committed:
Update advi example to show match of posteriors.

twiecki commented:
This is cool, here is an updated example that shows the NUTS posteriors together with the ADVI estimates: https://github.com/pymc-devs/pymc3/blob/2250386c826e8744e13654188074dd69dad829de/pymc3/examples/advi.ipynb

twiecki committed:
Add hierarchical GLM with ADVI. Seems to have problems.

twiecki committed:
Add comments to ADVI and change number of gradients to look at from 50 to 10 as in original paper.

twiecki commented:
I tried ADVI on a more elaborate example (the hierarchical GLM with radon). However, the results from NUTS and ADVI seem to disagree pretty significantly: https://github.com/pymc-devs/pymc3/blob/01312efb319f45ec81824f7eccfea69e7b5a7ba1/pymc3/examples/GLM-hierarchical-ADVI.ipynb

twiecki commented:
@akucukelbir I think this is a solid draft but there are still some discrepencies when running the same model in NUTS and ADVI (e.g.https://github.com/pymc-devs/pymc3/blob/980a96f392380cf0aca0382c4652856ada565340/pymc3/examples/GLM-hierarchical-ADVI.ipynb). Would you mind having a quick look at this PR? Specifically the algo here: https://github.com/pymc-devs/pymc3/blob/advi/pymc3/variational/ADVI.py

akucukelbir commented:
@twiecki cool! 

first a few questions, then a few comments. i'm not entirely sure what the plots are showing in that example. 

Q1. are there any transformations here? (i.e. are we estimating a variance term that needs to stay positive?) 

Q2. does this implement the mean-field version of ADVI? if so, there are endemic differences we expect between NUTS and ADVI, such as ADVI underestimating the marginal variances of the posterior.

Q3. how do we measure convergence here?

now some comments

C1. we've continued working on ADVI over in Stan. there's a major pull request that should get through soon (https://github.com/stan-dev/stan/pull/1590) that implements a few things of note: a) a new learning rate, b) a grid-search procedure for estimating `eta` in the learning rate equation

C2. initialization and convergence are two things that we always need to be careful about with variational techniques. for hierarchical models, initialization can make a big difference. during debugging, i might recommend trying a few random initializations when comparing to NUTS. in practice, you may even be able to see this in the value of the ELBO that ADVI reaches upon convergence.

i'll try to take a closer look at the code soon too!

twiecki commented:
> Q1. are there any transformations here? (i.e. are we estimating a variance term that needs to stay positive?)

Yes, all RVs are automatically transformed to be on the real line.

> Q2. does this implement the mean-field version of ADVI? if so, there are endemic differences we expect between NUTS and ADVI, such as ADVI underestimating the marginal variances of the posterior.

Yes it does. The first plot shows the NUTS posteriors as histograms and the ADVI approximation as a solid line (this is all in transformed space). As you can see, even the posterior means are quite different.

> Q3. how do we measure convergence here?

We don't. That might be the (or one) problem.

> C1. we've continued working on ADVI over in Stan. there's a major pull request that should get through soon (stan-dev/stan#1590) that implements a few things of note: a) a new learning rate, b) a grid-search procedure for estimating eta in the learning rate equation

Cool, hopefully not too hard to add here as well. What's the new learning rate?

> C2. initialization and convergence are two things that we always need to be careful about with variational techniques. for hierarchical models, initialization can make a big difference. during debugging, i might recommend trying a few random initializations when comparing to NUTS. in practice, you may even be able to see this in the value of the ELBO that ADVI reaches upon convergence.

OK, I'll look into that.

akucukelbir commented:
convergence: yeah, i'd plot the ELBO to at least visually get a sense of whether we've converged for this model.

learning rate: it's the old learning rate scaled by eta_0/sqrt(iteration_counter). as i mentioned we also do a gridsearch to figure out the ballpark of eta_0.

hope these help!

datnamer commented:
http://downhill.readthedocs.org/en/stable/

ADAGRAD and other optimization algos for theano

springcoil commented:
How is this looking? Is there any update on this feature. Variational Inference could be a big thing for PyMC3. Should I copy the files to a local repository and have a go at reproducing them and then offer some feedback?

datnamer commented:
@akucukelbir thanks for your help so far. Has Stan made any recent tweaks pymc should be aware of?

twiecki commented:
@springcoil I agree, your help would be greatly appreciated. I think we should start by returning and plotting the ELBO to make sure we're converging and have better traction on analyzing the result.

akucukelbir commented:
@datnamer my pleasure! nothing concrete since my last message on Aug 28, 2015. 

recently, we've begun investigating how well http://arxiv.org/pdf/1502.02846.pdf might work for ADVI. i'll keep you posted.

twiecki commented:
@akucukelbir Do you have a sense for how well ADVI works for hierarchical models? I just wonder if the divergent behavior I observed is perhaps just limitations of ADVI rather then a bug in our code.

akucukelbir commented:
@twiecki i am not aware of any endemic issues specifically related to applying ADVI to hierarchical models. (the usual variational inference effects still stand, such as underestimating marginal variances.)

we've used ADVI on a few hierarchical models and have seen good results. (however, i wouldn't be surprised to see mean-field ADVI fail for hierarchical models with strong posterior correlations.)

taku-y commented:
Hi PyMC3 developers, 

I'm really interested in this issue. I read ADVI.py and thought that the size of the random number variable 'n' in L79 is not consistent with that of input variables. I inserted the following code just after random sampling:

# Formula 6 in the paper
q = n \* exp(w) + u
f = theano.function([uw], q)
print(f(np.zeros(4)))
print(f(np.zeros(4)))
print(f(np.zeros(4)))
print(f(np.zeros(4)))

and executed the script in advi.py, then the output was

[-0.30273833 -0.30273833]
[-2.71521646 -2.71521646]
[ 0.15230697  0.15230697]
[-1.03554189 -1.03554189]

Always returns the same random value over all dimension. 

For the GLM-hierarchical ADVI example, I modified L79

n = r.normal(size=(175,)) # 175 dimensional spherical normal

(I tried to get the dimension from theano tensor variable, but failed with "MissingInputError". I couldnt solve it)

Also, I used smaller larning rate (-.001). The result seems reasonable for me excepting sigma_beta_interval. 

![index](https://cloud.githubusercontent.com/assets/4969395/13173514/40bfadf2-d741-11e5-8636-8133dcb7306e.png)

Hope this comment would help you with implementing advi in PyMC3.

twiecki commented:
@taku-y Thank you so much -- that seems like it might have been the missing key. I'll check how we can get the dimension.

twiecki committed:
BUG Fix advi issue of only sampling a single dimension.

twiecki commented:
@taku-y It's pretty easy to sample the correct dimension and the results look way better indeed:
![image](https://cloud.githubusercontent.com/assets/674200/13174499/4a3e029e-d704-11e5-96c6-dc5b3edf3769.png)

twiecki committed:
ENH Make learning rate and epsilon user-facing kwargs.

twiecki committed:
DOC Update advi example.

Commit:
Add ELBO.

Commit:
Number of MC samples for ELBO.

Commit:
Import theano.tensor as T.

Commit:
Compute ELBO with gradients.

Commit:
Remove unused imports.

fonnesbeck commented:
Pedantic housekeeping point: let's name the file `advi.py` (i.e. remove all caps) to be consistent with other modules.

datnamer commented:
Can advi do streaming or chunks in its current iteration? I think STAN is looking at something like that.

It would be cool to do database-> blaze-> numpy chunks -> pymc 3

fonnesbeck commented:
@datnamer Are you talking about streaming variational Bayes, along the lines of Broderick et al.? Or something more along the lines of expectation propagation, where the data are broken into parts and analyzed in parallel? In either case, I'm pretty sure the answer is "no", but the latter is certainly something I am interested in.

fonnesbeck commented:
I noticed that `advi` does not return values for `Deterministic` nodes. It should probably do this.

datnamer commented:
I was thinking of streaming variational bayes.

fonnesbeck commented on pymc3/variational/ADVI.py:
```diff
@@ -0,0 +1,118 @@
+
+__all__ = ['advi']
+
+def advi(vars=None, start=None, model=None, n=5000, learning_rate=-.001, epsilon=.1):
```
Learning rate should be a positive value, no?

Commit:
Added progress bar; changed learning rate to positive value

Commit:
Added progressbar to advi; changed learning rate to positive value

fonnesbeck commented:
Perhaps we should have `advi` return a named tuple?

twiecki commented:
actually I think streaming advi should be pretty simple by turning data
into a shared tensor and then updating it.
On Feb 21, 2016 19:13, "Chris Fonnesbeck" notifications@github.com wrote:

> Perhaps we should have advi return a named tuple?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/775#issuecomment-186876032.

Commit:
advi returns namedtuple

twiecki committed:
Rename ADVI.py to advi.py.

twiecki commented:
@datnamer Do you have a good use-case / example in mind for streaming variational inference?

twiecki commented:
Bayesian Neural networks would be a great use-case as is demonstrated here: http://arxiv.org/pdf/1206.6380.pdf

twiecki committed:
Add accurate elbo back in but make it optional.

Commit:
Replaced progress bar with ELBO monitoring

Commit:
Fixed typo; we need unit tests for this.

Commit:
Fixed output formatting

Commit:
Added additional line of ADVI progress output

fonnesbeck commented:
I notice that `advi` returns the transformed variables, rather than the original ones:

```
{'mu': array(-0.1774186062443244), 'sd_log': array(-0.11581644733052084)}
```

We need to return originals (in addition to the deterministics as previously noted).

fonnesbeck commented on pymc3/variational/advi.py:
```diff
@@ -0,0 +1,158 @@
+                                        inarray.tag.test_value])
+
+    # Naive Monte-Carlo
+    r = MRG_RandomStreams(seed=1)
```
Let's pass seed in as an argument.

datnamer commented:
@twiecki I have a dataset in mind but can't release it. There is a public dataset I've been interested in digging into, just have to find it. 

In general, I was thinking of any streaming or out of core usecase where the data is bigger than ram. There could be a blaze table interface to multiple backends that get converted into in memory chunks.

jsalvatier commented:
@datnamer can you describe the dataset/application even if you can't
release it?

On Mon, Feb 22, 2016 at 6:33 PM datnamer notifications@github.com wrote:

> @twiecki https://github.com/twiecki I have a dataset in mind but can't
> release it. There is a public dataset I've been interested in digging into,
> just have to find it.
> 
> In general, I was thinking of any streaming or out of core usecase where
> the data is bigger than ram. There could be a blaze table interface to
> multiple backends that get converted into in memory chunks.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pymc-devs/pymc3/pull/775#issuecomment-187490380.

fonnesbeck commented:
Changes required before we can merge:
- [x] add tests for ADVI
- [ ] return Deterministics (including untransformed variables) along with stochastics
- [ ] add warning/failure if discrete variables exist in model

Feel free to edit.

taku-y committed:
Support minibatch in advi.

Sample size correction.

Support minibatch in advi.

Remove assertion.

Support minibatch in advi.

Add an example notebook.

Fix advi function to be given minibatch RVs explicitly.

Remove notebooks for minibatch advi from version control.

Fix advi function to be given minibatch RVs explicitly.

Support minibatch in advi.

Fix a bug.

Support minibatch in advi.

taku-y committed:
Modify ELBO to include constant term.

taku-y committed:
Add a test for advi.

twiecki commented:
Seems like we only need to solve the last two items:
- [ ] add warning/failure if discrete variables exist in model

@taku-y Is that something you could do? Then adding an example for ADVI and mini-batch ADVI and we should be able to merge.

twiecki commented:
Actually we already have a simple ADVI example. So if you want to add your GMM mini-batch example we're pretty close!

fonnesbeck commented:
Related to one of the outstanding items, in general I don't think we should be returning auto-transformed variables at all, since they are just crutches for the sampler. I will create an issue for this.

twiecki commented:
Good point, but it might be required for e.g. PPC.

taku-y commented:
I will try to add error for discrete variables in the model. For the issue of deterministics, I'm not sure how should variables returned and its implementation, e.g., extracting deterministic variables from the model. 

Another issue is LogSumExp appeared in my notebook. It's very useful for advi to create mixture models, because advi cannot deal with discrete variables. But currently not included in Theano. So I think it should be included in PyMC3. What do you think about it?

twiecki commented:
> I will try to add error for discrete variables in the model. For the issue of deterministics, I'm not sure how should variables returned and its implementation, e.g., extracting deterministic variables from the model.

Yeah, I wouldn't bother with that. Just a simple check that raises an exception if the model contains discrete vars.

> Another issue is LogSumExp appeared in my notebook. It's very useful for advi to create mixture models, because advi cannot deal with discrete variables. But currently not included in Theano. So I think it should be included in PyMC3. What do you think about it?

I agree, lets add it to `pymc3.math`.

twiecki commented on pymc3/variational/advi.py:
```diff
@@ -0,0 +1,205 @@
+
+ADVIFit = namedtuple('ADVIFit', 'means, stds, elbo_vals')
+
+def advi(vars=None, start=None, model=None, n=5000, accurate_elbo=False, 
```
This needs a doc-string.

twiecki commented on pymc3/variational/advi.py:
```diff
@@ -0,0 +1,205 @@
+        w[var] = np.exp(w[var])
+    return ADVIFit(u, w, elbos)
+
+def advi_minibatch(vars=None, start=None, model=None, n=5000, n_mcsamples=1, 
```
doc-string.

twiecki commented:
The other thing we need are doc-strings, at least for the main entry functions. Best for all of them.

taku-y committed:
Add docstring.

taku-y committed:
Add logsumexp.

taku-y committed:
Add check for discrete variables.

twiecki commented:
Rebased. Waiting for travis to finish and will then merge.

twiecki commented:
Above, @akucukelbir wrote:

> learning rate: it's the old learning rate scaled by eta_0/sqrt(iteration_counter). as i mentioned we also do a gridsearch to figure out the ballpark of eta_0.

Seems quite easy to also add the scaled learning rate. Not sure how they do the grid-search.

akucukelbir commented:
hi guys. all this is looking great.

here is where/how we implemented the grid-search:

https://github.com/stan-dev/stan/blob/develop/src/stan/variational/advi.hpp#L162

it's a heuristic approach, so any other heuristic might also work here.

twiecki commented:
Thanks @akucukelbir. Does STAN do mini-batch ADVI?

I think we should go ahead and merge this as it works quite well. We can always improve it later.

Thanks to everyone here, this really was a team effort! @jsalvatier @fonnesbeck @akucukelbir and especially @taku-y.

akucukelbir commented:
yes, but it is unsupported.

https://github.com/stan-dev/stan/blob/adsvi/how_to_ADSVI.md

twiecki commented:
@taku-y Do you also want to add your GMM example NB?

taku-y commented:
I will rewrite NB because when using mini-batch we need to slightly modify the model. When finished, I will send a PR.

twiecki commented:
:+1:
